internal/httpcli/redis_logger_middleware.go

```diff
@@ -0,0 +1,166 @@
+	creatorStackFrame, _ := getFrames(4).Next()
+	return func(cli Doer) Doer {
+		return DoerFunc(func(req *http.Request) (*http.Response, error) {
+			start := time.Now()
+			resp, err := cli.Do(req)
+			duration := time.Since(start)
+			var requestBody []byte
+			if req != nil && req.Body != nil {
+				requestBody, _ = io.ReadAll(req.Body)
+			}
+
+			errorMessage := ""
+			if err != nil {
+				errorMessage = err.Error()
+			}
+			key := generateKey(time.Now())
+			callerStackFrame, _ := getFrames(4).Next() // Caller of the caller of redisLoggerMiddleware
+			logItem := types.OutboundRequestLogItem{
+				Key:                key,
+				StartedAt:          start,
+				Method:             req.Method,
+				URL:                req.URL.String(),
+				RequestHeaders:     removeSensitiveHeaders(req.Header),
+				RequestBody:        string(requestBody),
+				StatusCode:         int32(resp.StatusCode),
+				ResponseHeaders:    removeSensitiveHeaders(resp.Header),
+				Duration:           duration.Seconds(),
+				ErrorMessage:       errorMessage,
+				CreationStackFrame: formatStackFrame(creatorStackFrame),
+				CallStackFrame:     formatStackFrame(callerStackFrame),
+			}
+
+			logItemJson, jsonErr := json.Marshal(logItem)
+
+			if jsonErr != nil {
+				log.Error(jsonErr)
+			}
+
+			// Save new item
+			redisCache.Set(key, logItemJson)
+
+			deleteExcessItems()
+
+			return resp, err
+		})
+	}
+}
+
+func deleteExcessItems() {
+	deletionErr := redisCache.DeleteAllButLastN(nil, keyPrefix, OutboundRequestLogLimit())
+	if deletionErr != nil {
+		log.Error(deletionErr)
+	}
+}
+
+func generateKey(now time.Time) string {
+	return fmt.Sprintf("%s%s", keyPrefix, now.UTC().Format("2006-01-02T15_04_05.999999999"))
+}
+
+// GetAllOutboundRequestLogItemsAfter returns all outbound request log items after the given key,
+// in ascending order, trimmed to maximum {limit} items.
+// The given "lastKey" must contain `keyPrefix` as a prefix. Example: "outbound:2021-01-01T00_00_00.000000000".
+func GetAllOutboundRequestLogItemsAfter(lastKey *string, limit int) ([]*types.OutboundRequestLogItem, error) {
+	if limit == 0 {
+		return []*types.OutboundRequestLogItem{}, nil
+	}
+
+	rawItems, err := getAllValuesAfter(redisCache, keyPrefix, lastKey, limit)
+	if err != nil {
+		return nil, err
+	}
+
+	var items []*types.OutboundRequestLogItem
+	for _, rawItem := range rawItems {
+		var item types.OutboundRequestLogItem
+		err = json.Unmarshal(rawItem, &item)
+		if err != nil {
+			return nil, err
+		}
+		items = append(items, &item)
+	}
+	return items, nil
+}
+
+// getAllValuesAfter returns all items after the given key, in ascending order, trimmed to maximum {limit} items.
+// The given "lastKey" must contain `keyPrefix` as a prefix. Example: "outbound:2021-01-01T00_00_00.000000000".
+func getAllValuesAfter(redisCache *rcache.Cache, keyPrefix string, lastKey *string, limit int) ([][]byte, error) {
+	all, err := redisCache.ListKeys(nil, keyPrefix)
```

Comment:
The placeholder is `context.Background()` if you don't have a contex tied to, say, a user request.

```suggestion
	all, err := redisCache.ListKeys(context.Background(), keyPrefix)
```